File: sdks/node/src/registry-client.ts
import * as grpc from "@grpc/grpc-js";
import { RegistryServiceClient } from "../gen/coordin8/registry";
import type {
  Capability,
  RegistryEvent,
  RegistryEvent_EventType,
} from "../gen/coordin8/registry";

export interface TransportDescriptor {
  type: string;
  config: Record<string, string>;
}

export interface CapabilityRecord {
  capabilityId: string;
  interfaceName: string;
  attrs: Record<string, string>;
  transport?: TransportDescriptor;
}

export interface RegisterResult {
  capabilityId: string;
  leaseId: string;
  grantedTtlSeconds: number;
}

export interface RegistryEventRecord {
  type: "registered" | "expired" | "modified";
  capability?: CapabilityRecord;
}

export type Template = Record<string, string>;

function toRecord(c: Capability): CapabilityRecord {
  return {
    capabilityId: c.capabilityId,
    interfaceName: c.interface,
    attrs: c.attrs,
    transport: c.transport
      ? { type: c.transport.type, config: c.transport.config }
      : undefined,
  };
}

function eventTypeToString(t: RegistryEvent_EventType): "registered" | "expired" | "modified" {
  switch (t) {
    case 1: return "expired";
    case 2: return "modified";
    default: return "registered";
  }
}

export class RegistryClient {
  private readonly stub: RegistryServiceClient;

  constructor(channel: grpc.Channel) {
    this.stub = new RegistryServiceClient(
      "passthrough:///djinn",
      grpc.credentials.createInsecure(),
      { channelOverride: channel }
    );
  }

  /**
   * Register a service capability. Returns capability ID and lease info.
   * Pass capabilityId to update an existing registration in-place.
   */
  register(
    interfaceName: string,
    attrs: Record<string, string>,
    transport: TransportDescriptor | undefined,
    ttlSeconds: number,
    capabilityId?: string
  ): Promise<RegisterResult> {
    return new Promise((resolve, reject) => {
      this.stub.register(
        {
          interface: interfaceName,
          attrs,
          transport,
          ttlSeconds,
          capabilityId: capabilityId ?? "",
        },
        (err, res) => {
          if (err || !res) return reject(err);
          resolve({
            capabilityId: res.capabilityId,
            leaseId: res.lease?.leaseId ?? "",
            grantedTtlSeconds: res.lease?.ttlSeconds ?? 0,
          });
        }
      );
    });
  }

  /**
   * Modify attributes on an existing registration without re-registering.
   */
  modifyAttrs(
    capabilityId: string,
    addAttrs?: Record<string, string>,
    removeAttrs?: string[]
  ): Promise<CapabilityRecord> {
    return new Promise((resolve, reject) => {
      this.stub.modifyAttrs(
        {
          capabilityId,
          addAttrs: addAttrs ?? {},
          removeAttrs: removeAttrs ?? [],
        },
        (err, res) => {
          if (err || !res) return reject(err);
          resolve(toRecord(res));
        }
      );
    });
  }

  lookup(template: Template): Promise<CapabilityRecord | null> {
    return new Promise((resolve, reject) => {
      this.stub.lookup({ template }, (err, res) => {
        if (err) {
          if (err.code === grpc.status.NOT_FOUND) return resolve(null);
          return reject(err);
        }
        resolve(res ? toRecord(res) : null);
      });
    });
  }

  lookupAll(template: Template): Promise<CapabilityRecord[]> {
    return new Promise((resolve, reject) => {
      const results: CapabilityRecord[] = [];
      const stream = this.stub.lookupAll({ template });
      stream.on("data", (cap: Capability) => results.push(toRecord(cap)));
      stream.on("end", () => resolve(results));
      stream.on("error", reject);
    });
  }

  /**
   * Watch for registry events matching the template.
   * Returns a stream that emits RegistryEventRecord objects.
   * Call stream.cancel() to stop watching.
   */
  watch(
    template: Template,
    onEvent: (evt: RegistryEventRecord) => void,
    onError?: (err: Error) => void
  ): grpc.ClientReadableStream<RegistryEvent> {
    const stream = this.stub.watch({ template });
    stream.on("data", (evt: RegistryEvent) => {
      const rec: RegistryEventRecord = {
        type: eventTypeToString(evt.type),
        capability: evt.capability ? toRecord(evt.capability) : undefined,
      };
      onEvent(rec);
    });
    if (onError) {
      stream.on("error", onError);
    }
    return stream;
  }
}

File: sdks/java/src/main/java/io/coordin8/RegistryClient.java
package io.coordin8;

import coordin8.RegistryServiceGrpc;
import coordin8.Registry.RegisterRequest;
import coordin8.Registry.RegisterResponse;
import coordin8.Registry.ModifyAttrsRequest;
import coordin8.Registry.LookupRequest;
import coordin8.Registry.RegistryWatchRequest;
import coordin8.Registry.RegistryEvent;
import coordin8.Registry.Capability;
import io.grpc.ManagedChannel;
import io.grpc.StatusRuntimeException;
import io.grpc.stub.StreamObserver;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Consumer;

public class RegistryClient {

    private final RegistryServiceGrpc.RegistryServiceBlockingStub stub;
    private final RegistryServiceGrpc.RegistryServiceStub asyncStub;

    RegistryClient(ManagedChannel channel) {
        this.stub = RegistryServiceGrpc.newBlockingStub(channel);
        this.asyncStub = RegistryServiceGrpc.newStub(channel);
    }

    /**
     * Register a service capability with the Djinn.
     *
     * @param interfaceName the interface name (e.g., "Greeter")
     * @param attrs         service attributes
     * @param ttlSeconds    requested lease TTL in seconds
     * @param transport     optional transport descriptor (null if not needed)
     * @return the registration result with capability ID and lease
     */
    public RegisterResult register(String interfaceName, Map<String, String> attrs,
                                   long ttlSeconds, TransportDescriptor transport) {
        return register(interfaceName, attrs, ttlSeconds, transport, null);
    }

    /**
     * Register or re-register a service capability with the Djinn.
     *
     * @param interfaceName the interface name
     * @param attrs         service attributes
     * @param ttlSeconds    requested lease TTL in seconds
     * @param transport     optional transport descriptor
     * @param capabilityId  if non-null, updates an existing registration in-place
     * @return the registration result with capability ID and lease
     */
    public RegisterResult register(String interfaceName, Map<String, String> attrs,
                                   long ttlSeconds, TransportDescriptor transport,
                                   String capabilityId) {
        RegisterRequest.Builder req = RegisterRequest.newBuilder()
                .setInterface(interfaceName)
                .setTtlSeconds(ttlSeconds);
        if (attrs != null) {
            req.putAllAttrs(attrs);
        }
        if (transport != null) {
            req.setTransport(coordin8.Registry.TransportDescriptor.newBuilder()
                    .setType(transport.type())
                    .putAllConfig(transport.config())
                    .build());
        }
        if (capabilityId != null) {
            req.setCapabilityId(capabilityId);
        }

        RegisterResponse resp = stub.register(req.build());
        String leaseId = resp.hasLease() ? resp.getLease().getLeaseId() : null;
        long grantedTtl = resp.hasLease() ? resp.getLease().getTtlSeconds() : 0;
        return new RegisterResult(resp.getCapabilityId(), leaseId, grantedTtl);
    }

    /**
     * Modify attributes on an existing registration without re-registering.
     *
     * @param capabilityId the capability to modify
     * @param addAttrs     attributes to add or update (null to skip)
     * @param removeAttrs  attribute keys to remove (null to skip)
     * @return the updated capability
     */
    public CapabilityRecord modifyAttrs(String capabilityId,
                                        Map<String, String> addAttrs,
                                        List<String> removeAttrs) {
        ModifyAttrsRequest.Builder req = ModifyAttrsRequest.newBuilder()
                .setCapabilityId(capabilityId);
        if (addAttrs != null) {
            req.putAllAddAttrs(addAttrs);
        }
        if (removeAttrs != null) {
            req.addAllRemoveAttrs(removeAttrs);
        }
        Capability cap = stub.modifyAttrs(req.build());
        return toRecord(cap);
    }

    /**
     * Look up the first capability matching the template.
     * Template semantics: missing fields match anything,
     * values prefixed "contains:" or "starts_with:" use those operators.
     */
    public Optional<CapabilityRecord> lookup(Map<String, String> template) {
        try {
            Capability cap = stub.lookup(LookupRequest.newBuilder()
                    .putAllTemplate(template)
                    .build());
            return Optional.of(toRecord(cap));
        } catch (StatusRuntimeException e) {
            if (e.getStatus().getCode() == io.grpc.Status.Code.NOT_FOUND) {
                return Optional.empty();
            }
            throw e;
        }
    }

    /** Look up all capabilities matching the template. */
    public List<CapabilityRecord> lookupAll(Map<String, String> template) {
        Iterator<Capability> it = stub.lookupAll(LookupRequest.newBuilder()
                .putAllTemplate(template)
                .build());
        List<CapabilityRecord> results = new ArrayList<>();
        it.forEachRemaining(cap -> results.add(toRecord(cap)));
        return results;
    }

    /**
     * Watch for registry events matching the template.
     * The callback is invoked on a gRPC thread for each event.
     *
     * @param template  attribute template to filter events
     * @param onEvent   callback for each registry event
     * @param onError   callback for errors (stream will be closed)
     */
    public void watch(Map<String, String> template,
                      Consumer<RegistryEventRecord> onEvent,
                      Consumer<Throwable> onError) {
        RegistryWatchRequest req = RegistryWatchRequest.newBuilder()
                .putAllTemplate(template)
                .build();
        asyncStub.watch(req, new StreamObserver<RegistryEvent>() {
            @Override
            public void onNext(RegistryEvent evt) {
                String type = switch (evt.getType()) {
                    case EXPIRED -> "expired";
                    case MODIFIED -> "modified";
                    default -> "registered";
                };
                CapabilityRecord cap = evt.hasCapability() ? toRecord(evt.getCapability()) : null;
                onEvent.accept(new RegistryEventRecord(type, cap));
            }

            @Override
            public void onError(Throwable t) {
                onError.accept(t);
            }

            @Override
            public void onCompleted() {}
        });
    }

    static CapabilityRecord toRecord(Capability c) {
        TransportDescriptor transport = null;
        if (c.hasTransport()) {
            transport = new TransportDescriptor(
                    c.getTransport().getType(),
                    c.getTransport().getConfigMap()
            );
        }
        return new CapabilityRecord(
                c.getCapabilityId(),
                c.getInterface(),
                c.getAttrsMap(),
                transport
        );
    }

    public record TransportDescriptor(
            String type,
            Map<String, String> config
    ) {}

    public record CapabilityRecord(
            String capabilityId,
            String interfaceName,
            Map<String, String> attrs,
            TransportDescriptor transport
    ) {}

    public record RegisterResult(
            String capabilityId,
            String leaseId,
            long grantedTtlSeconds
    ) {}

    public record RegistryEventRecord(
            String type,
            CapabilityRecord capability
    ) {}
}
